File: common/src/main/java/generations/gg/generations/core/generationscore/common/network/packets/shop/S2COpenShopHandler.java
package generations.gg.generations.core.generationscore.common.network.packets.shop;

import generations.gg.generations.core.generationscore.common.client.screen.ShopScreen;
import generations.gg.generations.core.generationscore.common.network.ClientNetworkPacketHandler;
import generations.gg.generations.core.generationscore.common.world.entity.ShopOfferProvider;
import net.minecraft.client.Minecraft;

public class S2COpenShopHandler implements ClientNetworkPacketHandler<S2COpenShopPacket> {

    @Override
    public void handle(S2COpenShopPacket packet, Minecraft minecraft) {
        if (Minecraft.getInstance().level == null)
            return;

        if (packet.entityId() != -1 && Minecraft.getInstance().level.getEntity(packet.entityId()) instanceof ShopOfferProvider shopOfferProvider)
            Minecraft.getInstance().setScreen(new ShopScreen(shopOfferProvider));
        else if (Minecraft.getInstance().level.getBlockEntity(packet.pos()) instanceof ShopOfferProvider shopOfferProvider) {
            Minecraft.getInstance().setScreen(new ShopScreen(shopOfferProvider));
        }
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/network/packets/shop/S2COpenShopHandler.kt
package generations.gg.generations.core.generationscore.common.network.packets.shop

import generations.gg.generations.core.generationscore.common.client.screen.ShopScreen
import generations.gg.generations.core.generationscore.common.network.ClientNetworkPacketHandler
import generations.gg.generations.core.generationscore.common.world.entity.ShopOfferProvider
import net.minecraft.client.Minecraft

object S2COpenShopHandler : ClientNetworkPacketHandler<S2COpenShopPacket> {
    override fun handle(packet: S2COpenShopPacket, minecraft: Minecraft) {
        if (Minecraft.getInstance().level == null) return

        var entity = Minecraft.getInstance().level!!.getEntity(packet.entityId)

        if (packet.entityId != -1 && entity is ShopOfferProvider) Minecraft.getInstance().setScreen(ShopScreen(entity))
        else {
            var blockEntity = Minecraft.getInstance().level!!.getBlockEntity(packet.pos);

            if (blockEntity is ShopOfferProvider) {
                Minecraft.getInstance().setScreen(ShopScreen(blockEntity))
            }
        }
    }
}
